[MRG+2] Clean common tests

albertcthomas opened the pull request:
#### What does this implement/fix? Explain your changes.
This PR fixes typos and removes dupes in check_estimators.py.

It also adds `check_supervised_y_no_nan` to `yield_classifier_checks`. Maybe there was a good reason for not using it for classifiers. If this is the case I will remove the associated commit.
Note that some tests were then failing for `RidgeClassifier` and `RidgeClassifierCV` because there was no `check_X_y` in their `fit` methods before calling `LabelBinarizer`. However their `fit` methods then call `fit` from `_BaseRidge` and `_BaseRidgeCV` which has a `check_X_y`. The current implementation thus implies two calls of `check_X_y` for `RidgeClassifier` and `RidgeClassifierCV`.


#### Any other comments?
Other questions about common tests that came to me when implementing common tests for outlier detection estimators (PR #9270):
- is there a common test checking that `predict` returns int for classifiers?
- why isn't `check_estimator` used in the common tests?

<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
http://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

albertcthomas committed:
rm dupes

albertcthomas committed:
add check_supervised_y_no_nan in classifier checks: this implies changes for Ridge classifiers

albertcthomas committed:
fix docstrings/comments

jnothman commented:
Thanks. I've not yet reviewed.

I think check_estimator isn't used in common tests mostly for legacy
reasons: it was a late addition to provide a public interface to what was
being driven by our test suite. But perhaps that should change.

Predict need not return int for classifiers. It needs to return strings if
the input is strings. I'm pretty sure there are common tests for class
encoding behaviours of classifiers.

On 12 July 2017 at 21:23, Albert Thomas <notifications@github.com> wrote:

> What does this implement/fix? Explain your changes.
>
> This PR fixes typos and removes dupes in check_estimators.py.
>
> It also adds check_supervised_y_no_nan to yield_classifier_checks. Maybe
> there was a good reason for not using it for classifiers. If this is the
> case I will remove the associated commit.
> Note that some tests were then failing for RidgeClassifier and
> RidgeClassifierCV because there was no check_X_y in their fit methods
> before calling LabelBinarizer. However their fit methods then call fit
> from _BaseRidge and _BaseRidgeCV which has a check_X_y. The current
> implementation thus implies two calls of check_X_y for RidgeClassifier
> and RidgeClassifierCV.
> Any other comments?
>
> Other questions about common tests that came to me when implementing
> common tests for outlier detection estimators (PR #9270
> <https://github.com/scikit-learn/scikit-learn/pull/9270>):
>
>    - is there a common test checking that predict returns int for
>    classifiers?
>    - why isn't check_estimator used in the common tests?
>
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>   https://github.com/scikit-learn/scikit-learn/pull/9340
> Commit Summary
>
>    - rm dupes
>    - add check_supervised_y_no_nan in classifier checks: this implies
>    changes for Ridge classifiers
>    - fix docstrings/comments
>
> File Changes
>
>    - *M* sklearn/linear_model/ridge.py
>    <https://github.com/scikit-learn/scikit-learn/pull/9340/files#diff-0>
>    (6)
>    - *M* sklearn/utils/estimator_checks.py
>    <https://github.com/scikit-learn/scikit-learn/pull/9340/files#diff-1>
>    (12)
>
> Patch Links:
>
>    - https://github.com/scikit-learn/scikit-learn/pull/9340.patch
>    - https://github.com/scikit-learn/scikit-learn/pull/9340.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9340>, or mute the
> thread
> <https://github.com/notifications/unsubscribe-auth/AAEz65R56rW7TSfuoR0VpSQAw18-pJnxks5sNKzOgaJpZM4OVfLR>
> .
>

albertcthomas commented:
Thanks for the answers. I was wondering about predict returning int for classifiers as predict needs to return int for outlier detection estimators (see #8676). Thanks for the clarification.

I'll try do add check_estimator in common tests and see what failed.

We might also avoid calling check_X_y twice for Ridge classifiers.

jnothman commented:
In that sense, outlier detection is more akin to clustering: the labels
aren't provided by supervision, so the estimator needs to make something up
that in the case of clusters is just an arbitrary identifier, and in the
case of outlier detection is a meaningful binary distinction.

On 13 July 2017 at 01:51, Albert Thomas <notifications@github.com> wrote:

> Thanks for the answers. I was wondering about predict returning int for
> classifiers as predict needs to return int for outlier detection estimators
> (see #8676 <https://github.com/scikit-learn/scikit-learn/issues/8676>).
> Thanks for the clarification.
>
> I'll try do add check_estimator in common tests and see what failed.
>
> We might also avoid calling check_X_y twice for Ridge classifiers.
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/9340#issuecomment-314813203>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz69DaQkLSX4e00uZ3NOQxWGbCxcioks5sNOuPgaJpZM4OVfLR>
> .
>

albertcthomas commented:
I noticed that many changes are being done in PR #8022 so I think it's better to stick with a small cleaning in this PR.

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -640,7 +641,7 @@ def check_fit1d_1feature(name, estimator_orig):
 @ignore_warnings
 def check_fit1d_1sample(name, estimator_orig):
-    # check fitting 1d array with 1 feature
+    # check fitting 1d array with 1 sample
```
What does this currently test? It should test that a value error is raised. That's weird. And I guess for a 1d example, it's not specified whether its a sample or a feature and therefore we error now. The test was written before we deprecated this behavior

albertcthomas replied:
Indeed. I will replace the two fit_1d checks, `check_fit1d_1sample` and `check_fit1d_1feature`, with only one `check_fit1d` testing that a `ValueError` is raised.

amueller commented on sklearn/utils/estimator_checks.py:
```diff
@@ -597,7 +598,7 @@ def check_fit2d_1sample(name, estimator_orig):
+    # check by fitting a 2d array with only one feature
     rnd = np.random.RandomState(0)
     X = 3 * rnd.uniform(size=(10, 1))
     y = X[:, 0].astype(np.int)
```
again, what are we testing here? Some refactoring seems to have gone wrong? Or a deprecation cycle?

albertcthomas replied:
Fitting a 2d array with only 1 feature is either OK or raise a ValueError depending on the estimator. Should we remove this common check then?

We might want to remove `check_fit2d_1sample` as well which is either OK or raise a ValueError depending on the estimator.

jnothman replied:
Where should having a single feature result in a `ValueError`?

albertcthomas replied:
Using the data set generated in `check_fit2d_1feature` the following estimators return a `ValueError`:
- FeatureAgglomeration, ValueError('Found array with 1 feature(s) (shape=(10, 1)) while a minimum of 2 is required by FeatureAgglomeration.',)
- GraphLasso, ValueError('Found array with 1 feature(s) (shape=(10, 1)) while a minimum of 2 is required by GraphLasso.',)
- GraphLassoCV, ValueError('Found array with 1 feature(s) (shape=(10, 1)) while a minimum of 2 is required by GraphLassoCV.',)
- OrthogonalMatchingPursuitCV, ValueError('Found array with 1 feature(s) (shape=(10, 1)) while a minimum of 2 is required by OrthogonalMatchingPursuitCV.',)
- RANSACRegressor, ValueError('RANSAC could not find a valid consensus set. All `max_trials` iterations were skipped because each randomly chosen sub-sample failed the passing criteria. See estimator attributes for diagnostics (n_skips*).',)
- RandomizedLogisticRegression, ValueError('This solver needs samples of at least 2 classes in the data, but the data contains only one class: 1',)
- SelectKBest, ValueError("k should be >=0, <= n_features; got 10.Use k='all' to return all features.",)

If `k='all'`, `SelectKBest` does not raise any exception. For `RandomizedLogisticRegression`, the `ValueError` is raised because one of the random subsamples of the data set contains only 1 class. I don't know how this should be handled.

amueller replied:
I feel this is a left-over from the deprecation of accepting 1d X. I guess we should ensure that it either works or there is an informative error message. RANSAC and RandomizedLogisticRegression seem weird. SelectKBest could mention ``n_features`` but otherwise seems good?

albertcthomas replied:
For the specific purpose of this test, i.e accepting 1d X or returning informative error, we can set `k='all'` for `SelectKBest`, `sample_fraction=1` for `RandomizedLogisticRegression` and `residual_threshold=0.5` for `RANSACRegressor`.

albertcthomas committed:
FIX check fitting 1d X array raises error and FIX check fitting 2d array with only 1 feature either works or returns informative message

albertcthomas commented:
Thanks for the reviews. The two remaining points are:
- should I remove `chech_X_y` from `_BaseRidge` and `_BaseRidgeCV`? and make sure it is called by every subclass? `chech_X_y` is currently called twice when doing `RidgeClassifier.fit()` and `RidgeClassifierCV.fit()`
- About `check_fit2d_1sample`, should we test it either works or returns an informative message as we did for `check_fit2d_1feature`? This currently either works or returns a ValueError with different messages, depending on the estimator.

albertcthomas commented:
And maybe add some tests for some of the functions I changed in estimator_checks.py

albertcthomas commented:
I suggest leaving `chech_X_y` in `_BaseRidge` and `_BaseRidgeCV` as it will perform a check on the output of `LabelBinarizer` in `RidgeClassifier.fit`and `RidgeClassifierCV.fit` and thus perform a different check than the `check_X_y` in these `fit` methods.
@amueller what do you think about `check_fit2d_1sample`?

albertcthomas committed:
modify check_fit2d_1sample in common tests so that it checks fitting either works or returns an informative message

albertcthomas commented:
I modified `check_fit2d_1sample` so that it checks that fitting either works or returns an informative error message mentioning that X contains only one sample or y contains only one class. Some estimators were raising a ValueError because of arrays containing inf or nan (because of X containing only one sample). I added `ensure_min_samples=2` in `check_X_y` for those cases.

albertcthomas committed:
rm SpectralClustering case for the moment

jnothman commented on sklearn/cluster/mean_shift_.py:
```diff
@@ -68,7 +68,10 @@ def estimate_bandwidth(X, quantile=0.3, n_samples=None, random_state=0,
         X = X[idx]
-    nbrs = NearestNeighbors(n_neighbors=int(X.shape[0] * quantile),
+    n_neighbors = int(X.shape[0] * quantile)
+    if n_neighbors < 1:  # cannot fit NearestNeighbors with n_neighbors = 0
```
I think this deserves a unit test.

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -589,15 +592,30 @@ def check_fit2d_1sample(name, estimator_orig):
-    except ValueError:
-        pass
+    except ValueError as e:
+        if all(msg not in repr(e) for msg in msgs):
```
This can be done with `assert_raises_regexp` FWIW.

But this is quite a rag bag of error messages. I guess it's an improvement. But do we intend to cut down the number of acceptable ways for an estimator to state this? If so, maybe we should comment in here what the preferred message is.

albertcthomas replied:
We are testing that `fit` either works or returns an informative message. Can we really use `assert_raises_regexp` here?

agramfort replied:
I agree with @albertcthomas I don't think that assert_raises_regexp works here

jnothman replied:
okay

jnothman commented on sklearn/utils/estimator_checks.py:
```diff
@@ -608,42 +626,31 @@ def check_fit2d_1feature(name, estimator_orig):
-
     set_random_state(estimator, 1)
 
+    msgs = ["Found array with 1 feature(s)", "n_features = 1"]
```
I think it's fair to allow no spaces: n_features=1

jnothman reviewed (commented):
Otherwise this is looking good

albertcthomas commented:
thanks @jnothman ! `check_fit2d_1sample` is harder to fix than `check_fit2d_1feature` because of the different cases.
On second thoughts I'm not sure it was a good thing to add `ensure_min_samples=2` for estimators raising a ValueError about arrays containing inf or nan as this also depends on the chosen parameters of the estimator like the chosen `eigen_solver` in `TSNE` for instance. Besides Travis tests fail for this check because of estimators raising a ValueError about arrays containing inf or nan whereas the tests are OK on my machine. I should maybe add `array must not contain infs or NaNs` in the possible messages for now and consider this PR to be a small fix of this check. WDYT?

ngoix commented:
While cleaning check_estimator, i'm wondering if we really have to require all classifiers to support the multi-class setting. Do all the sklearn classifiers support it? Functions involving multi-class testing are `check_classifiers_classes` and `check_classifiers_train`.

albertcthomas commented:
From the multiclass section of the User guide: "All scikit-learn classifiers are capable of multiclass classification".

agramfort commented on sklearn/utils/estimator_checks.py:
```diff
@@ -221,10 +222,11 @@ def _yield_all_checks(name, estimator):
     yield check_fit2d_predict1d
-    yield check_fit2d_1sample
+    if name != 'GaussianProcess' and name != 'SpectralClustering':  # FIXME
+        # XXX GaussianProcess deprecated in 0.20
```
and SpectralClustering why?

agramfort commented:
besides @jnothman 's comments and mine LGTM. @albertcthomas can  you see why CIs are not happy?

albertcthomas commented:
thanks @agramfort.
- one of TSNE tests fails (`test_preserve_trustworthiness_approximalety`). I talked IRL with @ogrisel about this: this test is known to be unstable but this branch is older than the fix merged for TSNE in July so we might try a rebase...
- `check_fit2d_1sample` fails for `PLSRegression`, `PLSCanonical` and `CCA` because these estimators raise a ValueError with `array must not contain infs or NaNs`. The fix depends on how we want to address these messages: I can force them to return an informative message but I'm not sure it's the best solution, see my [comment above](https://github.com/scikit-learn/scikit-learn/pull/9340#issuecomment-327466356).

albertcthomas committed:
uniformize error messages for 1 sample case and fix SpectralClustering with ensure_min_samples=2

albertcthomas committed:
add unit test for mean_shift when n_samples * quantile < 1

jnothman commented on sklearn/manifold/t_sne.py:
```diff
@@ -706,6 +706,8 @@ def _fit(self, X, skip_num_points=0):
                             'memory. Otherwise consider dimensionality '
                             'reduction techniques (e.g. TruncatedSVD)')
+        if self.method == 'barnes_hut':
+            X = check_array(X, ensure_min_samples=2, dtype=np.float64)
```
why shouldn't barnes_hut support sparse matrices?

albertcthomas replied:
because of the previous if statement?

```python
if self.method == 'barnes_hut' and sp.issparse(X):
             raise TypeError('A sparse matrix was passed, but dense '
                             'data is required for method="barnes_hut"
```

jnothman commented:
The tSNE test failure shouldn't be due to not rebasing on master. The CIs automatically merge with master before building. It's strange that this test fails, as usually it passes CI, while for some users/platforms the test consistently fails. Now it seems to be consistently failing here. I don't get why that could happen.

albertcthomas committed:
FIX travis with ensure_min_samples=2 in _PLS

albertcthomas commented:
I simplified the list of error messages that we want to catch and I fixed PLS estimators. Let's wait for CI to end and see if the tSNE test is the only one failing...

albertcthomas committed:
try fix for failing tSNE test

albertcthomas commented:
CIs are green now. Thanks @jnothman, your comment about the tSNE test failure helped me find out why it was failing.

albertcthomas committed:
typos

agramfort commented on sklearn/gaussian_process/gpc.py:
```diff
@@ -595,8 +596,9 @@ def fit(self, X, y):
-                             "distinct classes. Only class %s present."
-                             % self.classes_[0])
+                             "distinct classes; got %d class."
+                             " Only class %s present."
```
remove this line its redundant

agramfort commented on sklearn/svm/base.py:
```diff
@@ -503,6 +503,7 @@ def _validate_targets(self, y):
         if len(cls) < 2:
             raise ValueError(
                 "The number of classes has to be greater than one; got %d"
+                " class"
```
can fit in one line

agramfort commented on sklearn/utils/estimator_checks.py:
```diff
@@ -608,42 +618,31 @@ def check_fit2d_1feature(name, estimator_orig):
     rnd = np.random.RandomState(0)
     X = 3 * rnd.uniform(size=(20))
-    y = np.array([1])
+    y = X.astype(np.int)
```
do you really need to change this? it seems that the logic of the test changed

albertcthomas replied:
there were previously two checks: `check_fit1d_1sample` and `check_fit1d_1feature`. But for a 1d array we should raise an error (because we don't know whether its a sample or a feature, see @amueller's [comment](https://github.com/scikit-learn/scikit-learn/pull/9340#discussion_r128802201)). I thus replaced these two checks by only one check `check_fit1d`.

agramfort commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1121,9 +1116,6 @@ def check_classifiers_train(name, classifier_orig):
                     assert_array_equal(np.argmax(decision, axis=1), y_pred)
 
-                # raises error on malformed input
-                assert_raises(ValueError,
```
can you tell me why this change is necessary?

albertcthomas replied:
because it's a duplicate of the next two lines

agramfort commented on sklearn/utils/estimator_checks.py:
```diff
@@ -1136,8 +1128,6 @@ def check_classifiers_train(name, classifier_orig):
             # check that probas for all classes sum to one
             assert_allclose(np.sum(y_prob, axis=1), np.ones(n_samples))
-            # raises error on malformed input
-            assert_raises(ValueError, classifier.predict_proba, X.T)
```
can you tell me why this change is necessary?

albertcthomas replied:
because it's a duplicate of the next two lines

agramfort commented:
besides LGTM thx @albertcthomas

albertcthomas committed:
take @agramfort's review into account

albertcthomas commented:
thanks for the review @agramfort

albertcthomas committed:
sc to fix string in gaussian_process

albertcthomas commented:
I just realize that (regarding `check_fit2d_1feature` and `check_fit2d_1sample') this PR also fixes #5380 which was being addressed in PR #5559 and basically made the same changes.

agramfort commented:
@jnothman @amueller ok for you?

one last review? maybe @lesteve or @glemaitre ?

jnothman commented on sklearn/gaussian_process/gpc.py:
```diff
@@ -189,8 +189,9 @@ def fit(self, X, y):
         elif self.classes_.size == 1:
-            raise ValueError("{0:s} requires 2 classes.".format(
-                self.__class__.__name__))
+            raise ValueError("{0:s} requires 2 classes; got {1:d} class"
```
If this PR is test-driven, why does this appear to have no test coverage?

albertcthomas replied:
That's because I must have changed all error messages in the file gpc.py related to the fact that there is only one class. However this is the error message of the estimator `_BinaryGaussianProcessClassifierLaplace` which is ignored in the common tests because of the name starts with `_`.

jnothman commented on sklearn/gaussian_process/gpc.py:
```diff
@@ -595,8 +596,8 @@ def fit(self, X, y):
         self.n_classes_ = self.classes_.size
         if self.n_classes_ == 1:
             raise ValueError("GaussianProcessClassifier requires 2 or more "
-                             "distinct classes. Only class %s present."
```
The previous message *is* more informative...

albertcthomas replied:
right, I added the present class again.

jnothman reviewed (commented):
LGTM

albertcthomas committed:
add the class that is present to preserve information of previous message in gpc.py

albertcthomas commented:
thanks for the review @jnothman !

agramfort commented:
Thx @albertcthomas

amueller commented:
Thanks!
